TestContext scope is not as documented, broken examples

### Describe the bug

https://vitest.dev/guide/test-context.html states that TestContext is setup before each test function and cleaned up after each test function, as documented in the sample code copy & pasted from there.

```ts
// my-test.ts
import { test } from 'vitest'

const todos = []
const archive = []

export const myTest = test.extend({
  todos: async ({ task }, use) => {
    // setup the fixture before each test function
    todos.push(1, 2, 3)

    // use the fixture value
    await use(todos)

    // cleanup the fixture after each test function
    todos.length = 0
  },
  archive
})
```

```ts
// xx.test.ts
import { expect } from 'vitest'
import { myTest } from './my-test.ts'

myTest('add items to todos', ({ todos }) => {
  expect(todos.length).toBe(3)

  todos.add(4) /// ADDED BY ME. This needs to be replaced by `todos.push(4)`
  expect(todos.length).toBe(4)
})

myTest('move items from todos to archive', ({ todos, archive }) => {
  expect(todos.length).toBe(3) /// ADDED BY ME. This fails because todos.length is 4. 
  expect(archive.length).toBe(0)

  archive.push(todos.pop())
  expect(todos.length).toBe(2)
  expect(archive.length).toBe(1)
})
```

After fixing the obvious error in the first test (todos.add is not defined, replaced by todos.push), the second test fails because the context is not reset between test executions as documented. Instead, a single TextContext is shared between both test executions and the cleanup is only executed after both tests are done.

### Reproduction

* Copy & paste the example files
* Run `vitest run xx` 
* Output:
```
 FAIL  src/t/xx.test.ts > move items from todos to archive
AssertionError: expected 4 to be 3 // Object.is equality

- Expected
+ Received

- 3
+ 4

 ❯ src/t/xx.test.ts:12:24
     10|
     11| myTest('move items from todos to archive', ({ todos, archive }) => {
     12|   expect(todos.length).toBe(3)
       |                        ^
     13|   expect(archive.length).toBe(0)
     14|

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed (1)
      Tests  1 failed | 1 passed (2)
   Start at  22:48:08
   Duration  190ms (transform 33ms, setup 0ms, collect 25ms, tests 6ms, environment 0ms, prepare 47ms)
```

### System Info

```shell
System:
    OS: macOS 13.6
    CPU: (10) arm64 Apple M1 Max
    Memory: 137.77 MB / 64.00 GB
    Shell: 3.5.1 - /opt/homebrew/bin/fish
  Binaries:
    Node: 20.8.1 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 10.1.0 - /opt/homebrew/bin/npm
    Watchman: 2023.10.09.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 118.0.5993.117
    Safari: 16.6
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/.eslintrc
+++ b/.eslintrc
@@ -14,6 +14,7 @@
         "paths": ["path"]
       }
     ],
+    "no-empty-pattern": "off",
     "import/no-named-as-default": "off",
     "no-cond-assign": "off"
   },

--- a/docs/guide/test-context.md
+++ b/docs/guide/test-context.md
@@ -86,7 +86,7 @@ const todos = []
 const archive = []
 
 export const myTest = test.extend({
-  todos: async ({ task }, use) => {
+  todos: async ({}, use) => {
     // setup the fixture before each test function
     todos.push(1, 2, 3)
 
@@ -104,12 +104,12 @@ Then we can import and use it.
 
 ```ts
 import { expect } from 'vitest'
-import { myTest } from './my-test.ts'
+import { myTest } from './my-test.js'
 
 myTest('add items to todos', ({ todos }) => {
   expect(todos.length).toBe(3)
 
-  todos.add(4)
+  todos.push(4)
   expect(todos.length).toBe(4)
 })
 

--- a/test/core/test/fixture-initialization.test.ts
+++ b/test/core/test/fixture-initialization.test.ts
@@ -144,4 +144,39 @@ describe('fixture initialization', () => {
       expect(b).toBe(2)
     })
   })
+
+  describe('fixture todos', () => {
+    const todos: number[] = []
+    const archive: number[] = []
+
+    const myTest = test.extend<{ todos: number[]; archive: number[] }>({
+      todos: async ({}, use) => {
+        // setup the fixture before each test function
+        todos.push(1, 2, 3)
+
+        // use the fixture value
+        await use(todos)
+
+        // cleanup the fixture after each test function
+        todos.length = 0
+      },
+      archive,
+    })
+
+    myTest('add items to todos', ({ todos }) => {
+      expect(todos.length).toBe(3)
+
+      todos.push(4)
+      expect(todos.length).toBe(4)
+    })
+
+    myTest('move items from todos to archive', ({ todos, archive }) => {
+      expect(todos.length).toBe(3)
+      expect(archive.length).toBe(0)
+
+      archive.push(todos.pop() as number)
+      expect(todos.length).toBe(2)
+      expect(archive.length).toBe(1)
+    })
+  })
 })

